oxenmq/base32z.h

```diff
@@ -74,44 +74,86 @@ static_assert(b32z_lut.from_b32z('w') == 20 && b32z_lut.from_b32z('T') == 17 &&
 
 } // namespace detail
 
-/// Converts bytes into a base32z encoded character sequence, writing them starting at `out`.
-/// Returns the final value of out (i.e. the iterator positioned just after the last written base32z
-/// character).
-template <typename InputIt, typename OutputIt>
-OutputIt to_base32z(InputIt begin, InputIt end, OutputIt out) {
-    static_assert(sizeof(decltype(*begin)) == 1, "to_base32z requires chars/bytes");
-    int bits = 0; // Tracks the number of unconsumed bits held in r, will always be in [0, 4]
-    std::uint_fast16_t r = 0;
-    while (begin != end) {
-        r = r << 8 | static_cast<unsigned char>(*begin++);
+/// Returns the number of characters required to encode a base32z string from the given number of bytes.
+inline constexpr size_t to_base32z_size(size_t byte_size) { return (byte_size*8 + 4) / 5; } // ⌈bits/5⌉ because 5 bits per byte
+/// Returns the (maximum) number of bytes required to decode a base32z string of the given size.
+inline constexpr size_t from_base32z_size(size_t b32z_size) { return b32z_size*5 / 8; } // ⌊bits/8⌋
 
-        // we just added 8 bits, so we can *always* consume 5 to produce one character, so (net) we
-        // are adding 3 bits.
-        bits += 3;
-        *out++ = detail::b32z_lut.to_b32z(r >> bits); // Right-shift off the bits we aren't consuming right now
+/// Iterable object for on-the-fly base32z encoding.  Used internally, but also particularly useful
+/// when converting from one encoding to another.
+template <typename InputIt>
+struct base32z_encoder final {
+private:
+    InputIt _it, _end;
+    static_assert(sizeof(decltype(*_it)) == 1, "base32z_encoder requires chars/bytes input iterator");
+    int bits; // Number of bits held in r; will always be >= 5 until we are at the end.
+    uint_fast16_t r;
+public:
+    using iterator_category = std::input_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = char;
+    using reference = value_type;
+    using pointer = void;
+    base32z_encoder(InputIt begin, InputIt end) : _it{std::move(begin)}, _end{std::move(end)} {
+        if (_it != _end) {
+            bits = 8;
+            r = static_cast<unsigned char>(*_it);
+        } else {
+            bits = 0;
```

Comment:
initialize `r` in the else?

Reply:
Why?  `r` is only used in incrementing and dereferencing, both of which are invalid when begin == end

Reply:
Fix pushed.

Reply:
... or maybe not because Apple.